pkg/ccl/sqlproxyccl/balancer/balancer.go

```diff
@@ -66,6 +143,84 @@ func (b *Balancer) randFloat32() float32 {
 	return b.mu.rng.Float32()
 }
 
+// processQueue runs on a background goroutine, and invokes TransferConnection
+// for each rebalance request.
+func (b *Balancer) processQueue(ctx context.Context) {
+	// acquire blocks until ctx has been cancelled, or it has acquired a
+	// process semaphore. This returns true if the acquire was successful, and
+	// false otherwise.
+	acquire := func() bool {
```

Comment:
nit: the DB uses https://github.com/marusama/semaphore as its semaphore implementation. Conveniently its Acquire method accepts a ctx.